### PR TITLE
docs(adrs): add correlation ids to FxA services

### PR DESCRIPTION
**Pull request by chenba**

### CODE REVIEW

**Review by vbudhram — COMMENTED** [review]
Thanks @chenba, I'll ponder on this a little more but it seems pretty straightforward way to solve the problem you are describing.

**Review by bbangert — COMMENTED** [review]
This looks good, it would be useful to mention that this is used by other distributed tracing tools such as Amazon X-Ray, and Google's Cloud Trace. Currently we use Amazon for most deployed services which add's a unique trace id to every incoming request, which we could/should retain on any additional internal requests. Google's LB's have a similar option for use with Cloud Trace, either/both of those we could use and retain for logging and any additional internal requests.

One difficulty of these unique request/trace/correlation id's, which I'm curious how we would deal with, is propagating them throughout the call stack such that they can be attached to the log output. In a threaded request environment this is simple because it can be a thread-local, in Node though, we'd have to pass the request or trace id down the call stack anywhere we might want a log message to include it. Was the thought here that we'd try and pass down this id everywhere we want to log?

File: docs/adr/0026-add-correlation-id.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+## Decision Drivers
+
+- Developer experience while tracing and debugging an issue across multiple FxA
+  microservices.
+- Ability to query relevant log entries for any given error.
+
+## Considered Options
+
+1. Add correlation ids to FxA services.
+1. Do nothing.
```
[the remark this answers is not in the document]
Comment (chenba) [reply docs/adr/0026-add-correlation-id.md:29]: The correlation id is simpler.  Our flow id is an HMAC, which allows us to use it from our frontends.  We don't need the authentication property of the id for the correlation ids.

I'll add it as one of the options so we can discuss the pros and cons.  Thanks!

File: docs/adr/0026-add-correlation-id.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+Have FxA services generate or accept correlation ids in their request headers
+and use the ids in logs and error responses.
+
+When a client (RP, web frontend, etc.) request arrives on a service, the
+service must generate a header named "x-correlation-id" with a [UUID version
+4][uuid-rfc] as the value and append it to the request's headers.
+
+For inter-service requests, the client should include the correlation ID as the
+"x-correlation-id" header in its request. The receiving service should accept
+the header and use the value as its correlation ID. If the sending service
```
[the remark this answers is not in the document]
Comment (chenba) [reply docs/adr/0026-add-correlation-id.md:52]: No.  I will make that clear.  Thanks!

**Comment by chenba** [comment]
> 
> 
> it would be useful to mention that this is used by other distributed tracing tools such as Amazon X-Ray, and Google's Cloud Trace. ...

I'll add that to the pro column.  Thanks!  And I'll check out the docs on the tracing ids from the cloud provider.

> 
> One difficulty of these unique request/trace/correlation id's, which I'm curious how we would deal with, is propagating them ... Was the thought here that we'd try and pass down this id everywhere we want to log?

Sadly, yes, hence one of items in the cons list.  We could experiment a bit with zone.js here I suppose.

**Comment by bbangert** [comment]
From my reading, zones.js might have some issues in large apps. `cls-hooked` however seems to use an official Nodejs async hook API and may work better in this case. We might want to try a POC to see how that works out in practice.

**Comment by chenba** [comment]
Given our current engineering capacity and product priorities, I'm quite uncertain when we'll get to tackle this.  Moreover, in the time since I drafted the ADR, I did not notice any engineers voicing a need for better traceability through an id like the one suggested in the ADR.  I will close the PR for now.  We can revisit when we have the need and the capacity to implement.

@bbangert @vbudhram Thank you for your input!

**Pull request closed without merging**